cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -643,6 +643,10 @@ Declaration Declaration::Sequence(std::vector<Declaration> decls) {
   return out;
 }
 
+bool Declaration::IsValid(ExecFactoryRegistry* registry) const {
+  return registry->GetFactory(this->factory_name).ok() && this->options != nullptr;
```

Comment:
```suggestion
  return !this->factory_name.empty() && this->options != nullptr;
```

I think the earlier advice I provided was slightly off.  I think, all we need to do here is distinguish between "default-constructed invalid declarations" and "properly constructed declarations".  Actual validation (e.g. hash-map lookup in the factory) doesn't really belong in this method as it should be really fast.